src/descriptor/mod.rs

```diff
@@ -87,10 +87,71 @@ pub struct DescriptorSinglePub {
     pub key: bitcoin::PublicKey,
 }
 
+#[derive(Debug)]
+pub struct DescriptorSinglePriv {
+    pub origin: Option<bip32::KeySource>,
+    pub key: bitcoin::PrivateKey,
+}
+
+#[derive(Debug)]
+pub enum DescriptorSecretKey {
+    SinglePriv(DescriptorSinglePriv),
+    XPrv(DescriptorXKey<bip32::ExtendedPrivKey>),
+}
+
+impl fmt::Display for DescriptorSecretKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &DescriptorSecretKey::SinglePriv(ref sk) => {
+                maybe_fmt_master_id(f, &sk.origin)?;
+                sk.key.fmt(f)?;
+                Ok(())
+            }
+            &DescriptorSecretKey::XPrv(ref xprv) => {
+                maybe_fmt_master_id(f, &xprv.origin)?;
+                xprv.xkey.fmt(f)?;
+                fmt_derivation_path(f, &xprv.derivation_path)?;
+                if xprv.is_wildcard {
+                    write!(f, "/*")?;
+                }
+                Ok(())
+            }
+        }
+    }
+}
+
+/// Trait for "extended key" types like `xpub` and `xprv`. Used internally to generalize parsing and
+/// handling of `bip32::ExtendedPubKey` and `bip32::ExtendedPrivKey`.
+pub trait InnerXKey: fmt::Display + str::FromStr {
+    fn xkey_fingerprint(&self) -> bip32::Fingerprint;
+    fn can_derive_hardened() -> bool;
+}
+
+impl InnerXKey for bip32::ExtendedPubKey {
+    fn xkey_fingerprint(&self) -> bip32::Fingerprint {
+        self.fingerprint()
+    }
+
+    fn can_derive_hardened() -> bool {
+        false
+    }
+}
+
+impl InnerXKey for bip32::ExtendedPrivKey {
+    fn xkey_fingerprint(&self) -> bip32::Fingerprint {
+        self.fingerprint(&secp256k1::Secp256k1::signing_only())
```

Comment:
:skull_and_crossbones: we definitely should not be creating a one-off secp context for this